clients/git/client.go

```diff
@@ -0,0 +1,209 @@
+// Copyright 2021 OpenSSF Scorecard Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package git defines helper functions for clients.RepoClient interface.
+package git
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"sync"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	cp "github.com/otiai10/copy"
+
+	"github.com/ossf/scorecard/v4/clients"
+)
+
+const repoDir = "repo*"
+
+var (
+	errNilCommitFound = errors.New("nil commit found")
+	errEmptyQuery     = errors.New("query is empty")
+)
+
+type Client struct {
+	gitRepo        *git.Repository
+	worktree       *git.Worktree
+	listCommits    *sync.Once
+	tempDir        string
+	errListCommits error
+	commits        []clients.Commit
+	commitDepth    int
+}
+
+func (c *Client) InitRepo(uri, commitSHA string, commitDepth int) error {
+	// cleanup previous state, if any.
+	c.Close()
+	c.listCommits = new(sync.Once)
+	c.commits = nil
+
+	// init
+	c.commitDepth = commitDepth
+	tempDir, err := os.MkdirTemp("", repoDir)
+	if err != nil {
+		return fmt.Errorf("os.MkdirTemp: %w", err)
+	}
+
+	// git clone
+	const filePrefix = "file://"
+	if strings.HasPrefix(uri, filePrefix) {
+		if err := cp.Copy(strings.TrimPrefix(uri, filePrefix), tempDir); err != nil {
+			return fmt.Errorf("cp.Copy: %w", err)
+		}
+		c.gitRepo, err = git.PlainOpen(tempDir)
+		if err != nil {
+			return fmt.Errorf("git.PlainOpen: %w", err)
+		}
+	} else {
+		c.gitRepo, err = git.PlainClone(tempDir, false /*isBare*/, &git.CloneOptions{
```

Comment:
I see comments that the e2e tests take a while, does adding a clone depth help?

https://pkg.go.dev/github.com/go-git/go-git/v5#CloneOptions

Reply:
Well problem with clone depth here is that we won't be able to support `--commit` option.

Reply:
True. We could always optimize for the common case though and add a clone depth if `commitSHA == clients.HeadSHA`

Reply:
I would prefer to leave this as is for now without over optimizing for performance. One reason being, I want to try and test out the proposal [here](https://github.com/ossf/scorecard/issues/2016#issuecomment-1468381435) which assumes Scorecard has access to all commit history. If clone depth is really a performance blocker, we can reconsider. Let me know if that works.

Reply:
SGTM